src/codegen.jl

```diff
@@ -355,20 +339,33 @@ end
 
 function debug_actions(machine::Machine)
     actions = Set{Symbol}()
-    for trans in values(machine.transitions)
-        for (_, as) in values(trans)
-            union!(actions, as)
+    for s in traverse(machine.start)
+        for (e, t) in s.edges
+            union!(actions, a.name for a in e.actions)
         end
     end
     for as in values(machine.eof_actions)
-        union!(actions, as)
+        union!(actions, a.name for a in as)
     end
     function log_expr(name)
         return :(push!(logger, $(QuoteNode(name))))
     end
     return Dict(name => log_expr(name) for name in actions)
 end
 
+# Generic foldr.
+function foldr(op::Function, x0, xs)
```

Comment:
makes sense - would it be worth giving this a more specific name or type signature just so it's clear it's a package-local method?